edge/consumer.go

```diff
@@ -0,0 +1,69 @@
+package edge
+
+type Consumer struct {
+	edge Edge
+	r    Receiver
+}
+
+func NewConsumer(edge Edge, r Receiver) *Consumer {
+	return &Consumer{
+		edge: edge,
+		r:    r,
+	}
+}
+
+func (ec *Consumer) Run() error {
+	for msg, ok := ec.edge.Next(); ok; msg, ok = ec.edge.Next() {
+		switch typ := msg.Type(); typ {
+		case BeginBatch:
```

Comment:
I played around with this idea, and it currently can only remove two cases from the switch statement and has only one level of nesting. It doesn't seem like much of a gain for the complexity it adds. And I don't foresee many nested messages types emerging.

When we circle back to performance test this we can reconsider.